numpy/core/tests/test_scalar_methods.py

```diff
@@ -201,3 +202,227 @@ def test_bit_count(self):
+        self.identical(ftype.fromhex('nan'), np.nan)
+        self.identical(ftype.fromhex('+NaN'), np.nan)
+        self.identical(ftype.fromhex('-NaN'), np.nan)
+        self.identical(ftype.fromhex('-nAN'), np.nan)
+
+        # variations in input format
+        self.identical(ftype.fromhex('1'), 1.0)
+        self.identical(ftype.fromhex('+1'), 1.0)
+        self.identical(ftype.fromhex('1.'), 1.0)
+        self.identical(ftype.fromhex('1.0'), 1.0)
+        self.identical(ftype.fromhex('1.0p0'), 1.0)
+        self.identical(ftype.fromhex('01'), 1.0)
+        self.identical(ftype.fromhex('01.'), 1.0)
+        self.identical(ftype.fromhex('0x1'), 1.0)
+        self.identical(ftype.fromhex('0x1.'), 1.0)
+        self.identical(ftype.fromhex('0x1.0'), 1.0)
+        self.identical(ftype.fromhex('+0x1.0'), 1.0)
+        self.identical(ftype.fromhex('0x1p0'), 1.0)
+        self.identical(ftype.fromhex('0X1p0'), 1.0)
+        self.identical(ftype.fromhex('0X1P0'), 1.0)
+        self.identical(ftype.fromhex('0x1P0'), 1.0)
+        self.identical(ftype.fromhex('0x1.p0'), 1.0)
+        self.identical(ftype.fromhex('0x1.0p0'), 1.0)
+        self.identical(ftype.fromhex('+0x1p0'), 1.0)
+        self.identical(ftype.fromhex('0x01p0'), 1.0)
+        self.identical(ftype.fromhex('0x1p00'), 1.0)
+        self.identical(ftype.fromhex(' 0x1p0 '), 1.0)
+        self.identical(ftype.fromhex('\n 0x1p0'), 1.0)
+        self.identical(ftype.fromhex('0x1p0 \t'), 1.0)
+        self.identical(ftype.fromhex('0xap0'), 10.0)
+        self.identical(ftype.fromhex('0xAp0'), 10.0)
+        self.identical(ftype.fromhex('0xaP0'), 10.0)
+        self.identical(ftype.fromhex('0xAP0'), 10.0)
+        self.identical(ftype.fromhex('0xbep0'), 190.0)
+        self.identical(ftype.fromhex('0xBep0'), 190.0)
+        self.identical(ftype.fromhex('0xbEp0'), 190.0)
+        self.identical(ftype.fromhex('0XBE0P-4'), 190.0)
+        self.identical(ftype.fromhex('0xBEp0'), 190.0)
+        self.identical(ftype.fromhex('0xB.Ep4'), 190.0)
+        self.identical(ftype.fromhex('0x.BEp8'), 190.0)
+        self.identical(ftype.fromhex('0x.0BEp12'), 190.0)
+
+        # moving the point around
+
+        # results that should overflow...
+        large_values = [
+            f"-0x1p{MAXEXP}", f"0x1p+{MAXEXP + 1}", f"+0X1p{MAXEXP + 5}",
+            f"-0x1p+{MAXEXP + 50}", f"0X1p123456789123456789",
+            f"+0X.8p+{MAXEXP + 1}", f"+0x0.8p{MAXEXP + 1}",
+            f"-0x0.4p{MAXEXP + 2}", f"0X2p+{MAXEXP - 1}",
+            f"0x2.p{MAXEXP - 1}", f"-0x2.0p+{MAXEXP - 1}",
+            f"+0X4p+{MAXEXP - 2}",
+            f"0x1.{'f' * (NMANT // 4 + 1)}p+{MAXEXP - 1}",
+            # for below case, last `f` is just to be sure
+            # for 128. It can also be `9` for 64, etc
+            f"-0X1.{'f' * (NMANT // 4)}fp{MAXEXP - 1}",
+            f"+0x3.{'f' * (NMANT // 4 + 1)}p{MAXEXP - 2}",
+            # XXX How are we getting to 970?
+            # f"0x3fffffffffffffp+{MAXEXP - NMANT}",
```

Comment:
Hi @mdickinson, can you please help me understand how we arrive at `970` for `64` bits? It exactly gives `1.7976931348623158079e+308` when max is `1.7976931348623157e+308`. Please ignore `{MAXEXP - NMANT}`, it was just a placeholder.

This is in reference to https://github.com/python/cpython/blob/e18d81569fa0564f3bc7bcfd2fce26ec91ba0a6e/Lib/test/test_float.py#L1332

Also, thanks for the extensive testsuite, was able to find a myriad of bugs :).

Reply:
`0x3fffffffffffffp+970` represents the smallest positive value that overflows rather than rounding to a finite number.

As to derivation: the maximum finite representable value is `2**DBL_MAX_EXP - 2**(DBL_MAX_EXP - DBL_MANT_DIG)` (in terms of the constants defined in the C standard). However, this isn't the boundary at which we start to overflow: that boundary is halfway between the above value and `2**DBL_MAX_EXP`, so it's `2**DBL_MAX_EXP - 2**(DBL_MAX_EXP - DBL_MANT_DIG - 1)`, which assuming IEEE 754 binary64, is `2**1024 - 2**970`, or `(2**54 - 1) * 2**970`.

For IEEE 754 single-precision (float32), the corresponding boundary would be `2**128 - 2**103`, or `0x1ffffffp+103`. For `float16`, it would be `2**16 - 2**4` (`0xfffp+4`):

A bit off-topic, but there's something a bit funny going on with int-to-float32 conversions in NumPy. I suspect a double-rounding through float64 first.

```python
>>> np.float32(2**128 - 2**103)  # ok: gives infinity, as expected
inf
>>> np.float32(2**128 - 2**103 - 1)  # bad: expect a finite result, get infinity instead
inf
```

It looks as though the actual boundary is at `2**128 - 2**103 - 2**74` instead of `2**128 - 2**103`:

```
>>> np.float32(2**128 - 2**103 - 2**74)
inf
>>> np.float32(2**128 - 2**103 - 2**74 - 1)
3.4028235e+38
```

Reply:
Thanks a lot for the explanation @mdickinson! Will incorporate the same.